v3/tasks/20250214-v3-protocol-fee-controller-v2/test/task.fork.ts

```diff
@@ -0,0 +1,193 @@
+
+    const authorizerTask = new Task('20210418-authorizer', TaskMode.READ_ONLY, getForkedNetwork(hre));
+    authorizer = await authorizerTask.deployedInstance('Authorizer');
+
+    const weightedTask = new Task('20241205-v3-weighted-pool', TaskMode.READ_ONLY, getForkedNetwork(hre));
+    factory = await weightedTask.deployedInstance(FACTORY_CONTRACT_NAME);
+
+    input = task.input() as ProtocolFeeControllerDeployment;
+    admin = await getSigner();
+  });
+
+  before('setup contracts and parameters', async () => {
+    tokenConfig = [
+      {
+        token: input.WETH,
+        tokenType: 0,
+        rateProvider: ZERO_ADDRESS,
+        paysYieldFees: false,
+      },
+      {
+        token: input.BAL,
+        tokenType: 0,
+        rateProvider: ZERO_ADDRESS,
+        paysYieldFees: false,
+      },
+    ].sort(function (a, b) {
+      return a.token.toLowerCase().localeCompare(b.token.toLowerCase());
+    });
+  });
+
+  it('deploys contract', async () => {
+    expect(await feeController.vault()).to.eq(input.Vault);
+  });
+
+  it('grants permissions', async () => {
+    const govMultisig = await impersonate(GOV_MULTISIG, fp(100));
+
+    // Allow admin to upgrade the protocol fee controller to the version with events.
+    await authorizer
+      .connect(govMultisig)
+      .grantRole(
+        await vaultAdmin.getActionId(vaultAdmin.interface.getSighash('setProtocolFeeController')),
+        admin.address
+      );
+
+    // Allow admin to set global fees, so that pools will be created with non-zero initial percentages.
+    // No standard factories allow turning off fees, but that's fine. We're just testing that the event comes through.
+    await authorizer
+      .connect(govMultisig)
+      .grantRole(
+        await feeController.getActionId(feeController.interface.getSighash('setGlobalProtocolSwapFeePercentage')),
+        admin.address
+      );
+
+    await authorizer
+      .connect(govMultisig)
+      .grantRole(
+        await feeController.getActionId(feeController.interface.getSighash('setGlobalProtocolYieldFeePercentage')),
+        admin.address
+      );
+  });
+
+  it('replaces the old fee controller', async () => {
+    const vaultAsExtension = vaultExtension.attach(vault.address);
+    const oldFeeControllerAddress = await vaultAsExtension.getProtocolFeeController();
+
+    const vaultAsAdmin = vaultAdmin.attach(vault.address);
+    await vaultAsAdmin.connect(admin).setProtocolFeeController(feeController.address);
+
+    const newFeeControllerAddress = await vaultAsExtension.getProtocolFeeController();
+    expect(newFeeControllerAddress).not.to.eq(oldFeeControllerAddress);
+    expect(newFeeControllerAddress).to.eq(feeController.address);
+  });
+
+  it('sets non-zero global fees', async () => {
+    await feeController.connect(admin).setGlobalProtocolSwapFeePercentage(GLOBAL_SWAP_FEE_PERCENTAGE);
+    await feeController.connect(admin).setGlobalProtocolYieldFeePercentage(GLOBAL_YIELD_FEE_PERCENTAGE);
```

Comment:
Hm, this is a good point.

Not in scope for this PR, but do you think we should do a script to do the controller upgrade + set these settings all in the same call?

Reply:
Since they're non-zero and presumably want to be continued, yes that would be more efficient. It could read the previous values and pass them through to the new one, so that we don't need to worry about anything hard-coded changing.